### PR TITLE
Remove Individual Data Type Caps in Per-shard Buffering for Remote Write

### LINKED ISSUE #8788: investigate per shard buffering improvements for remote write
Currently a shard allocates a buffer of `prompb.TimeSeries`, which is extended to be slightly longer than `maxSamplesPerSend` if sending of exemplars is enabled. We also allocate buffers of `prompb.Sample` and `prompb.Exemplar`, which are used to actually store the data of a sample or exemplar. The `prompb.TimeSeries` buffer then just references elements in the other two buffers before generating the actual `WriteRequest`. This saves us some allocations.

However, even though the the `prompb.TimeSeries` buffer can hold `maxSamplesPerSend*1.1` `promob.TimeSeries`, we only check if we have `maxSamplesPerSend` OR `maxExemplars` before building a write request. The `maxSamplesPerSend` defaults to `100` and `maxExemplars` is set to `int(math.Max(1, float64(maxSamplesPerSend/10)))`, or 10 in the case of default config. This means the worst case scenario is that we could end up sending write requests that only have 10 samples and 10 exemplars, potentially impacting throughput. 

We should investigate a way to buffer a reasonable amount of data points, regardless of type, and not having a max of one kind lead to reduced throughput.

### PR DESCRIPTION
**Pull request by LeviHarrison**
This PR consolidates the `sampleBuffer` and `exemplarBuffer` buffers into the `pendingData` buffer, removing the individual cap for each respective data type in favor of having a cap for all data points combined.

Fixes #8788

### CODE REVIEW

**Comment by roidelapluie** [comment]
Thanks for this. @cstyan would you have some time to review it?

@LeviHarrison did you run any benchmarks on this?

**Comment by LeviHarrison** [comment]
Not yet, I probably should.

**Comment by cstyan** [comment]
I should be able to review next week, end of this week is a holiday here. Benchmarks with memory stats would be good :)

**Comment by LeviHarrison** [comment]
I set up a benchmark using Grafana's TNS demo, so plenty of exemplars. I took this screenshot after it had been running for about 10 minutes.

test-dashboard

**Comment by LeviHarrison** [comment]
After about 20 minutes:

benchmark-2

**Comment by LeviHarrison** [comment]
Speaking of benchmarking... https://github.com/prometheus/test-infra/issues/249

File: storage/remote/queue_manager.go [hunk 1]
```diff
@@ -1021,19 +1021,15 @@ func (s *shards) runShard(ctx context.Context, shardID int, queue chan interface
...
-		pendingData    []prompb.TimeSeries
-		exemplarBuffer [][]prompb.Exemplar
+		buf         []byte
+		pendingData []prompb.TimeSeries
 	)
-	totalPending := max
 	if s.qm.sendExemplars {
-		exemplarBuffer = allocateExemplarBuffer(maxExemplars)
-		totalPending += maxExemplars
+		max += maxExemplars
```
Comment (cstyan) [inline storage/remote/queue_manager.go:1029]: I'm wondering if we should make this just `max += max*0.1` or have the increased amount be configurable. The other option is to just document a suggestion of: 
> increase your `max_samples_per_send by X% if you have send_exemplars enabled`

I'm leaning towards the last option, as we have a lot of configurable knobs already.

@csmarchbanks thoughts?

**Review by cstyan — COMMENTED** [review]
Thanks for taking this on @LeviHarrison, looking good so far. The benchmark screenshots look good if we can see those results consistently; the same allocations and overall memory usage but better pending samples throughput.

It would be nice if we could benchmark this in a more automated fashion, especially for any more potential future work. Maybe `BenchmarkSampleDelivery` could be extended to also send exemplars? 

Dave Cheney's post here has some good content on benchmarks: https://dave.cheney.net/high-performance-go-workshop/gophercon-2019.html#profiling_benchmarks

File: storage/remote/queue_manager.go (hunk 1 above)
Comment (csmarchbanks) [reply storage/remote/queue_manager.go:1029]: Hmm, I do agree that I do not want additional configuration options. I would prefer not to document that suggestion yet as I wonder if we could do something better than counting samples and exemplars separately. I would think that much of the time when sending exemplars they will be added close to the sample for that same series and we could combine exemplars and samples into the same series.

TLDR I would say just do `max += max*0.1` if exemplars are enabled for now.

Comment (csmarchbanks) [reply storage/remote/queue_manager.go:1029]: In other words, will we even need max += after https://github.com/prometheus/prometheus/pull/9075 is merged?

Comment (cstyan) [reply storage/remote/queue_manager.go:1029]: > In other words, will we even need max += after #9075 is merged?

Exactly. No, I don't think so.

Comment (LeviHarrison) [reply storage/remote/queue_manager.go:1029]: I'll make it `max += max*0.1`, although I wonder if we should document that just so users know why extra samples are being sent. Then again it seems that it probably won't make it past #9075 so maybe it doesn't matter.

**Comment by LeviHarrison** [comment]
`// TODO(cstyan): Casting this many times smells, also we could get index out of bounds issues here.`

I removed this comment from where `maxExemplars` was assigned, is it still valid?

**Comment by LeviHarrison** [comment]
@cstyan I'll take a look at expanding the benchmarks. I think that will probably end up as a separate PR thought.

File: storage/remote/queue_manager.go [hunk 2]
```diff
@@ -1078,23 +1071,21 @@ func (s *shards) runShard(ctx context.Context, shardID int, queue chan interface
 			// stop reading from the queue. This makes it safe to reference pendingSamples by index.
 			switch d := sample.(type) {
 			case writeSample:
-				sampleBuffer[nPendingSamples][0] = d.sample
 				pendingData[nPending].Labels = labelsToLabelsProto(d.seriesLabels, pendingData[nPending].Labels)
-				pendingData[nPending].Samples = sampleBuffer[nPendingSamples]
+				pendingData[nPending].Samples = []prompb.Sample{d.sample}
```
Comment (csmarchbanks) [inline storage/remote/queue_manager.go:1075]: By creating a new slice of samples I fear that we are reducing some of the allocation optimizations made in https://github.com/prometheus/prometheus/pull/5849. I wonder if we could still pre-allocate samples with len=0 and cap=1, then either set the value or truncate the slice as we use them?

Here is a benchmark you can use to see these concerns:
```
name               old time/op    new time/op    delta
SampleDelivery-16     15.5s ± 2%     16.9s ± 1%   +8.72%  (p=0.001 n=7+6)

name               old alloc/op   new alloc/op   delta
SampleDelivery-16    20.5GB ± 0%    21.7GB ± 0%   +5.85%  (p=0.001 n=7+7)

name               old allocs/op  new allocs/op  delta
SampleDelivery-16      176M ± 0%      201M ± 0%  +14.20%  (p=0.001 n=7+7)
```

**Review by csmarchbanks — COMMENTED** [review]
Generally I think this is close, but left one performance concern, and I think a solution for it.

File: storage/remote/queue_manager.go (hunk 2 above)
Comment (LeviHarrison) [reply storage/remote/queue_manager.go:1075]: I think what you're saying is that when we allocate `pendingData` we should also allocate len=0 series and exemplar slices to pre-populate the fields. The concern there was we'd over-allocate exemplars, but I think that if an allocation now happens for every single new sample, the initial penalty is well worth it.

Comment (csmarchbanks) [reply storage/remote/queue_manager.go:1075]: Yep, if someone has send exemplars turned on chances are they have some exemplars so eventually all of those would be allocated anyway, and it is much better than creating garbage that needs to be cleaned up for every sample.

Comment (LeviHarrison) [reply storage/remote/queue_manager.go:1075]: Actually, after trying to implement this I don't think it will work. Before when there were two separate slices they would only get sent up the index they had been filled, eliminated the possibility of a sample from the last cycle being sent again. But, because now each slice can hold _either_ a series or exemplar, the sample from the last cycle won't get wiped out if they aren't of the same type as the new sample. The only way no samples get re-sent is if the slice is filled in the exact same order every single time, which isn't guaranteed.

Comment (csmarchbanks) [reply storage/remote/queue_manager.go:1075]: Hmm, yeah this may add a fair bit more complexity. My thought is something like:
```
pendingData[nPending].Samples = pendingData[nPending].Samples[:0]
pendingData[nPending].Exemplars = pendingData[nPending].Exemplars[:0]
switch d := sample.(type) {
case writeSample:
    pendingData[nPending].Samples = append(pendingData[nPending].Samples, d.sample)
case writeExemplar:
    pendingData[nPending].Exemplars = append(pendingData[nPending].Exemplars, d.exemplar)
}
```
Do you think that would work? You might be able to use some ifs to avoid the truncations of the slices, but I don't know if it would help with performance at all.

Comment (LeviHarrison) [reply storage/remote/queue_manager.go]: I'm not sure, let's give it a try. Just pushed the change, tests pass. I'm not sure what benchmark you were using or else I would run it myself.

Comment (csmarchbanks) [reply storage/remote/queue_manager.go]: I have been running:
```
go test -bench=BenchmarkSampleDelivery -benchmem -run=djfkasldf ./storage/remote
```
for the benchmark. Here are the results for before this change, without preallocation, and with preallocation:
```
benchstat before.txt no_preallocation.txt preallocation.txt
name \ time/op     before.txt   no_preallocation.txt  preallocation.txt
SampleDelivery-16   15.5s ± 2%            16.9s ± 1%         15.1s ± 1%

name \ alloc/op    before.txt   no_preallocation.txt  preallocation.txt
SampleDelivery-16  20.5GB ± 0%           21.7GB ± 0%        20.5GB ± 0%

name \ allocs/op   before.txt   no_preallocation.txt  preallocation.txt
SampleDelivery-16    176M ± 0%             201M ± 0%          176M ± 0%
```

So looks like we are back to the same amount of allocations now (yay!) and might even be a couple percent faster compared to before this PR.

**Review by csmarchbanks — APPROVED** [review]
:+1:

File: storage/remote/queue_manager.go [hunk 3]
```diff
@@ -1016,24 +1016,20 @@ func (s *shards) runShard(ctx context.Context, shardID int, queue chan interface
...
-		exemplarBuffer = allocateExemplarBuffer(maxExemplars)
-		totalPending += maxExemplars
+		max += int(float64(max) * 0.1)
 	}
 
-	pendingData = make([]prompb.TimeSeries, totalPending)
+	var pendingData = make([]prompb.TimeSeries, max)
+	for i := range pendingData {
+		pendingData[i].Samples = []prompb.Sample{{}}
+		pendingData[i].Exemplars = []prompb.Exemplar{{}}
```
Comment (csmarchbanks) [inline storage/remote/queue_manager.go:1031]: I just realized: I guess we actually only need to allocate this if `s.qm.sendExemplars`? We would probably also have to protect line 1074 with an if as well at that point.

Comment (LeviHarrison) [reply storage/remote/queue_manager.go:1031]: True

**Comment by LeviHarrison** [comment]
We could also go the approach of:

```go
var reset func()
if !s.qm.sendExemplars {
  reset = func() {
    pendingData[nPending].Samples = pendingData[nPending].Samples[:0]
  }
} else {
  reset = func() {
    pendingData[nPending].Samples = pendingData[nPending].Samples[:0]
    pendingData[nPending].Exemplars = pendingData[nPending].Exemplars[:0]
  }
}
```

might be cleaner, maybe even better for performance, IDK

**Review by csmarchbanks — APPROVED** [review]
Approving the small new changes as well. @cstyan any other concerns on this one, otherwise lets merge it?

**Comment by cstyan** [comment]
:+1: lgtm

**Comment by LeviHarrison** [comment]
Thanks for all your time reviewing!
